### ISSUE TITLE
Parallelism correctness for RL training: logits parity & convergence across TP/PP/CP layouts

### ISSUE THREAD

**Issue by gcanlin** [vllm-project/vime#46]
# TP / PP correctness — coverage audit for `test_qwen3_0.6B_parallel_check.py`

## What this test is supposed to verify

In RL fine-tuning (GRPO / PPO), the **rollout** phase generates trajectories with vLLM, and the **training** phase recomputes log-probabilities with Megatron to assemble the policy-gradient loss. Both runtimes execute the same model, but under completely different conditions:

- Different inference / training kernels (vLLM vs Megatron)
- Different parallelism layouts (vLLM-side TP only; Megatron-side TP × PP × CP × SP × DP × EP)
- Weight transfer in-between (colocate: in-process IPC; non-colocate: NCCL broadcast)

If any one of these paths is numerically wrong, the importance-sampling ratio `π_new / π_old` becomes biased, advantages get mis-weighted, and the model **silently trains to the wrong objective** — no crash, no obvious loss spike.

`test_qwen3_0.6B_parallel_check.py` is the parity check that catches this class of bug. Its design:

1. **Phase A — baseline.** Run a vanilla DP-only training step. Persist:
   - `data-{i}.pt` — the rollout output (tokens + sampling-time log-probs)
   - `grad_norms-{i}.pt` — the per-step gradient norms from the training pass
2. **Phase B — sweep.** Replay the **same** `data-{i}.pt` under a grid of `(TP, PP, CP)` configurations on `{2, 4, 8}` GPUs. Each run must reproduce the baseline `grad_norms-{i}.pt` (within FP tolerance).
3. **Outer loop** runs Phase A + B twice — first with token-mean loss reduction, then with `--calculate-per-token-loss`.

Comparing `grad_norm` (a single scalar that depends on both forward logits and backward gradient assembly) against a frozen rollout input isolates any divergence to the training-side parallelism implementation. That is the entire point of the test.

The vLLM side is held constant throughout: `--rollout-num-gpus-per-engine 2 --rollout-num-gpus 8` → 4 engines, each at TP=2. Sweeping happens **only on the Megatron training side**.

## Full case matrix

Cases are derived from the four-level loop at [tests/test_qwen3_0.6B_parallel_check.py#L91-L133](https://github.com/THUDM/slime/blob/main/tests/test_qwen3_0.6B_parallel_check.py#L91-L133).

Constant across all rows: `--colocate`, vLLM TP=2 with 4 engines, `--global-batch-size 32`, `--rollout-batch-size 4`, `--n-samples-per-prompt 8`, `--num-rollout 1`. Sweep rows additionally pass `--sequence-parallel --use-dynamic-batch-size --max-tokens-per-gpu 8192`.

Two outer iterations `i ∈ {0, 1}`: iteration 0 uses the default token-mean loss; iteration 1 appends `--calculate-per-token-loss`. The table below lists each case once; both iterations execute the full set.

`✓` = covered by `test_qwen3_0.6B_parallel_check_2gpu.py`.

### Baseline runs

| # | Actor GPUs | TP | PP | CP | Note | 2-GPU variant |
|---|------------|----|----|----|------|---------------|
| B | 8 | 1 (default) | 1 | 1 | DP=8, writes `data-{i}.pt` + `grad_norms-{i}.pt` | (replaced by DP=2 baseline) |

### Sweep — `num_gpus = 8`

| # | TP | PP | CP | Effective DP | 2-GPU variant |
|---|----|----|----|--------------|---------------|
| 1 | 1 | 1 | 1 | 8 |   |
| 2 | 1 | 1 | 2 | 4 |   |
| 3 | 1 | 1 | 4 | 2 |   |
| 4 | 1 | 1 | 8 | 1 |   |
| 5 | 1 | 2 | 1 | 4 |   |
| 6 | 1 | 2 | 2 | 2 |   |
| 7 | 1 | 2 | 4 | 1 |   |
| 8 | 1 | 4 | 1 | 2 |   |

### Sweep — `num_gpus = 4`

| # | TP | PP | CP | Effective DP | 2-GPU variant |
|---|----|----|----|--------------|---------------|
| 9  | 1 | 1 | 1 | 4 |   |
| 10 | 1 | 1 | 2 | 2 |   |
| 11 | 1 | 1 | 4 | 1 |   |
| 12 | 1 | 2 | 1 | 2 |   |
| 13 | 1 | 2 | 2 | 1 |   |
| 14 | 2 | 1 | 1 | 2 |   |

### Sweep — `num_gpus = 2`

| # | TP | PP | CP | Effective DP | 2-GPU variant |
|---|----|----|----|--------------|---------------|
| 15 | 1 | 1 | 1 | 2 | ✓ |
| 16 | 1 | 1 | 2 | 1 | ✓ |
| 17 | 1 | 2 | 1 | 1 | ✓ |

### Totals

| | per outer iteration | total (× 2 outer iterations) |
|---|---|---|
| `test_qwen3_0.6B_parallel_check.py` | 1 baseline + 17 sweep = **18** | **36** |
| `test_qwen3_0.6B_parallel_check_2gpu.py` | 1 baseline + 3 sweep = **4** | **8** |

### 1. Sweep loop is buggy — high-TP configurations are silently skipped

[tests/test_qwen3_0.6B_parallel_check.py#L105-L115](https://github.com/THUDM/slime/blob/main/tests/test_qwen3_0.6B_parallel_check.py#L105-L115):

```python
for num_gpus in [8, 4, 2]:
    remaining_gpus = num_gpus
    for tp_size in [1, 2, 4, 8]:
        remaining_gpus /= tp_size            # accumulates across tp iterations
        for pp_size in [1, 2, 4]:
            if remaining_gpus < pp_size:
                continue
            remaining_gpus /= pp_size        # accumulates across pp iterations
            for cp_size in [1, 2, 4, 8]:
                if remaining_gpus < cp_size:
                    continue
                ...
```

`remaining_gpus` is a shared mutable counter across all inner iterations; the apparent intent is `tp_size * pp_size * cp_size <= num_gpus`. With the current code, by the time the loop reaches `tp_size=2`, `remaining_gpus` has already been divided through earlier `pp` / `cp` iterations and is too small to enter any inner block.

**Consequence — configurations that look like they are in the matrix but never actually execute:**

- `num_gpus=8`: `(tp=2,pp=1,cp=1)`, `(tp=2,pp=1,cp=2)`, `(tp=2,pp=1,cp=4)`, `(tp=2,pp=2,cp=1)`, `(tp=2,pp=2,cp=2)`, `(tp=4,pp=1,cp=1)`, `(tp=4,pp=1,cp=2)`, `(tp=4,pp=2,cp=1)`, `(tp=8,pp=1,cp=1)`
- `num_gpus=4`: `(tp=2,pp=1,cp=2)`, `(tp=2,pp=2,cp=1)`, `(tp=4,pp=1,cp=1)`
- `num_gpus=2`: `(tp=2,pp=1,cp=1)`

That is, **every configuration with `TP ≥ 2` plus another nontrivial axis is missing** — exactly the configurations the test is supposed to validate.

Fix: replace the cumulative `remaining_gpus` arithmetic with a single guard:

```python
for tp_size in [1, 2, 4, 8]:
    for pp_size in [1, 2, 4]:
        for cp_size in [1, 2, 4, 8]:
            if tp_size * pp_size * cp_size > num_gpus:
                continue
            ...
```

### 2. Non-colocate is not exercised

`misc_args` hard-codes `--colocate` at [tests/test_qwen3_0.6B_parallel_check.py#L77](https://github.com/THUDM/slime/blob/main/tests/test_qwen3_0.6B_parallel_check.py#L77). The colocate and non-colocate paths use different weight-transfer code (`update_weight_from_tensor.py` vs `update_weight_from_distributed.py`) and very different memory-management flows. The acceptance criterion calls out both modes, but only one is currently executed.

Fix: wrap the existing two-iteration `i` loop in an outer mode loop, and adjust `--rollout-num-gpus` / `--actor-num-gpus-per-node` so they sum to the available GPU budget in non-colocate mode.

### 3. Larger / production-like parallelism layouts are absent

The current grid stops at 8 GPUs and at `tp ≤ 8, pp ≤ 4, cp ≤ 8`. Cross-node PP, large CP, and combined TP+PP+CP layouts that production runs would actually use are out of scope. This is reasonable for a fast CI test, but should be acknowledged.

## Action items

- [ ] Fix the `remaining_gpus` loop bug — recover the `(TP ≥ 2, …)` configurations that are currently silently skipped
- [ ] Extend the test to cover the non-colocate path (acceptance criterion)
- [ ] Confirm the recovered `TP ≥ 2` cases actually pass — they have not been exercised before, so this is also where new bugs are most likely to surface
- [ ] (Optional) Promote `test_qwen3_0.6B_parallel_check_2gpu.py` into a smoke variant for low-resource runners, with a documented relationship to the full test

## Reproduce

Full test (8 GPUs):

```bash
cd /path/to/slime
python tests/ci/gpu_lock_exec.py --count 8 -- \
    python tests/test_qwen3_0.6B_parallel_check.py
```

Reduced 2-GPU variant (4 cases × 2 outer iterations = 8 runs):

```bash
cd /path/to/slime
CUDA_VISIBLE_DEVICES=4,5 \
    python tests/test_qwen3_0.6B_parallel_check_2gpu.py
```

**Comment by gcanlin** [comment]
@aoshen02 I tested these cases based on #18. And it works well.

**Comment by gcanlin** [comment]
# TP convergence — coverage audit for Qwen3-4B / Qwen3-30B-A3B under TP=2/4/8

## What this verifies

The core model tests (`test_qwen3_4B_ppo.py`, `test_qwen3_30B_A3B.py`, …) run a short end-to-end RL loop (rollout → reward → PPO/GRPO train → weight sync). They hard-code a single tensor-parallel layout. This task runs each test across the **TP = {2, 4, 8}** matrix and confirms training converges identically regardless of how the Megatron training side is sharded.

"Converge end-to-end" in this harness means:

1. All Ray jobs reach `succeeded`.
2. The `--ci-test` sanity asserts hold (log_probs / entropy / KL in expected ranges; no NaN/Inf).
3. The critic value-loss (PPO) / policy loss decreases consistently and the **trajectory matches across TP layouts within noise** — TP/PP/CP are mathematical identities, so a correct implementation must reproduce the same learning curve.
4. No crash, OOM, or weight-sync failure across the multi-rollout run.

Note the rollout (vLLM) side stays at `tensor_parallel_size=2` throughout (`--rollout-num-gpus-per-engine 2`); only the Megatron **training** TP varies. The two are decoupled.

## Results — Qwen3-4B (`test_qwen3_4B_ppo.py`, 8×A100-80GB, colocate, PPO + critic)

All three layouts: 8 GPUs, PP=1, `num_rollout=3`, `global_batch_size=32`.

| Layout | Ray job | `--ci-test` asserts | critic value_loss (step 0→1→2) | actor grad_norm | eval/aime24 (base) |
|--------|---------|---------------------|--------------------------------|-----------------|--------------------|
| **TP=2** CP=2 DP=2 | ✅ succeeded | pass | 8.99 → 6.41 → **3.93** | 0.43–0.50 | 0.667 |
| **TP=4** CP=1 DP=2 | ✅ succeeded | pass | 9.01 → 6.40 → **3.79** | 0.46–0.52 | 0.567 |
| **TP=8** CP=1 DP=1 | ✅ succeeded | pass | 9.29 → 6.48 → **4.08** | 0.52–0.52 | 0.567 |

**Verdict: PASS.** The critic value-loss curves overlap across all three layouts (8.99/9.01/9.29 → ~6.4 → 3.8–4.1), which is the meaningful TP-invariance signal — the critic is a freshly-initialized value head, so a matched learning trajectory across shardings demonstrates forward/backward/optimizer equivalence. Actor grad_norm is stable at 0.4–0.5 with no explosion.

Caveats:
- `eval/aime24` differences (0.667 vs 0.567) are **noise, not a convergence signal**: `--eval-interval 20` with `num_rollout=3` means eval runs once before any meaningful training, so it measures the *base model*. With `n-samples-per-eval-prompt=1` over 30 problems and differing vLLM seeds (1235/1236/1237), a 0.1 gap = 3 problems = sampling variance.
- 3 rollouts is a smoke-level run. It proves "trains cleanly and consistently under each TP", not "reaches a target reward".

## Coverage matrix

| Model | Test | TP=2 | TP=4 | TP=8 |
|-------|------|------|------|------|
| Qwen3-4B | `test_qwen3_4B_ppo.py` | ✅ | ✅ | ✅ |
| Qwen3-4B | `test_qwen3_4B_ckpt.py` | ❌ | ❌ | ❌ |
| Qwen3-30B-A3B | `test_qwen3_30B_A3B.py` | ❌ | ❌ | ❌ |
| Qwen3-30B-A3B | `test_qwen3_30B_A3B_r3.py` | ❌ | ❌ | ❌ |

Legend: ✅ run and passing · ❌ not yet run.

## Gaps / remaining work

- [ ] **Qwen3-30B-A3B TP matrix** — the MoE model is the harder half (EP=8, alltoall dispatch, FP8 path). None of TP=2/4/8 has been run. TP=2 (DP=4) is the OOM risk; TP=8 (DP=1) is the most memory-comfortable. This is the load-bearing part of the acceptance.
- [ ] **`test_qwen3_4B_ckpt.py` under TP matrix** — exercises the checkpoint save/load path (Megatron torch_dist resharding across TP), which `test_qwen3_4B_ppo.py` does not cover.
- [ ] **PP dimension** — every layout above is PP=1. If the acceptance intends "parallelism matrix" broadly, PP≥2 layouts (e.g. TP=2,PP=2) are untested for these models.
- [ ] **Real convergence curve** — current runs are 3-rollout smoke tests. A longer run (higher `num_rollout`, `--eval-interval` < `num_rollout`, `SLIME_TEST_ENABLE_EVAL=1`) is needed to confirm reward actually improves and that the improvement curve is TP-invariant, not just the loss.

## Reproduce

The tests were parameterized to read `SLIME_TP` / `SLIME_PP` / `SLIME_CP` from the environment (defaults preserve the original hard-coded layout).

Qwen3-4B (done):

```bash
cd /path/to/slime
for cfg in "2:2" "4:1" "8:1"; do
  TP=${cfg%:*}; CP=${cfg#*:}
  SLIME_TP=$TP SLIME_PP=1 SLIME_CP=$CP \
    python tests/test_qwen3_4B_ppo.py 2>&1 | tee conv_4B_tp${TP}cp${CP}.log
done
```

Qwen3-30B-A3B (pending):

```bash
cd /path/to/slime
# TP=4 (baseline) first, TP=2 last (highest OOM risk)
for cfg in "4:2" "8:1" "2:2"; do
  TP=${cfg%:*}; CP=${cfg#*:}
  SLIME_TP=$TP SLIME_PP=1 SLIME_CP=$CP \
    python tests/test_qwen3_30B_A3B.py 2>&1 | tee conv_30B_tp${TP}cp${CP}.log
done
```
